### ISSUE TITLE
Path F: Direct multi-horizon forecasting — eliminate autoregressive inference

### ISSUE THREAD

**Issue by Polichinel** [views-platform/views-hydranet#41]
## Sequence Position

> **This is Step 4b of 6** in the zero-inflation instability remediation sequence: **A → E → D/B → C/F**
>
> **Depends on:** Outcomes of #36 (Path A), #37 (Path E), #38 (Path D), #39 (Path B) — this path is a **strategic alternative** to pursue only if long-horizon quality remains poor despite teacher forcing (Path E).
> **Parallel with:** #40 (Path C: ZINB) — these combine naturally (as demonstrated by DynAttn) but are independently valuable.
>
> **This is an architecture redesign, not a fix.** It eliminates the autoregressive loop entirely, removing the class of instability problems by construction. Estimated effort: months. Pursue only with clear evidence that autoregressive instability is a fundamental bottleneck.

---

### The fundamental limitation of autoregression

The current model predicts one timestep at a time, feeding predictions back as input:

```
input(t) → predict(t+1) → feed back → predict(t+2) → ... → predict(t+36)
```

This creates two inherent problems:

1. **Error compounding:** Prediction errors at step t become input at step t+1, biasing predictions further. Over 36 steps, even small per-step errors accumulate into significant drift.
2. **Exposure bias:** The model is trained on ground-truth inputs but runs on its own predictions during inference. Teacher forcing (Path E, #37) mitigates but cannot fully eliminate this distributional shift.

Path E (GTF/scheduled sampling) reduces these problems but cannot eliminate them — they are structural properties of autoregressive inference. As long as predictions feed back, errors compound.

### The alternative: predict all horizons at once

Direct multi-horizon forecasting predicts all 36 future timesteps from a single forward pass:

```
input(t) → predict(t+1, t+2, ..., t+36) simultaneously
```

No feedback loop. No error compounding. No exposure bias. The prediction at t+36 is conditioned on the same input as t+1, not on 35 intermediate predictions.

### Domain validation: DynAttn

Iacus et al. (2025) demonstrate that direct multi-horizon forecasting works on the exact same VIEWS/PRIO-Grid conflict data, achieving R² = 0.897 even at horizon 12 — a quality level that autoregressive models typically cannot sustain due to compounding. Their DynAttn architecture uses:

- Transformer-based dynamic attention over spatial and temporal features
- ZINB likelihood for zero-inflation (see Path C, #40)
- Direct prediction of all horizons H=1,...,12 in a single forward pass

The key result: **prediction quality degrades gracefully with horizon** (R² from 0.955 at H=1 to 0.897 at H=12), rather than the exponential degradation typical of autoregressive models under distribution shift.

---

## Literature Support

| Paper | Authors | Year | Key finding | Link |
|-------|---------|------|-------------|------|
| DynAttn | Iacus et al. | 2025 | Direct multi-horizon forecasting on VIEWS/PRIO-Grid. ZINB + dynamic attention. R²=0.955 (H=1) → 0.897 (H=12). Substantially outperforms autoregressive competitors. | [arXiv](https://arxiv.org/abs/2512.21435) |
| Scheduled Sampling | Bengio et al. | 2015 | Motivates teacher forcing mitigation but notes that the fundamental train/inference mismatch in AR models cannot be fully resolved by curriculum approaches. | [NeurIPS](https://papers.nips.cc/paper/2015/hash/e995f98d56967d946471af29d7bf99f1-Abstract.html) |
| GTF | Hess et al. | 2023 | Bounds gradients for chaotic AR dynamics, but acknowledges the approach is a mitigation, not an elimination, of the exposure bias problem. | [ICML](https://proceedings.mlr.press/v202/hess23a.html) |

---

## Architectural Implications

This is the highest-disruption path. Almost every component of the current pipeline would change.

### What changes

| Component | Current (AR) | Direct multi-horizon |
|-----------|-------------|---------------------|
| **Temporal model** | ConvLSTM — sequential state updates | Temporal encoder (transformer, TCN, or temporal conv) — parallel processing |
| **Decoder** | Single-step output [B, C, H, W] | Multi-step output [B, T, C, H, W] or T separate heads |
| **Inference loop** | 36 sequential forward passes with feedback | Single forward pass |
| **Hidden state** | Maintained and updated across steps | Consumed by encoder, not propagated to decoder |
| **Training** | Per-window sequential processing | Per-window parallel processing (all horizons at once) |
| **Memory** | O(1) per step (stream predictions) | O(T) for all horizons simultaneously |
| **Temporal consistency** | Implicit via recurrence | Must be explicitly regularized |
| **Uncertainty** | MC-Dropout over 36 sequential passes | MC-Dropout over single pass (faster but potentially less diverse) |

### What stays the same

- Spatial model (UNet encoder-decoder) — the spatial processing is independent of temporal architecture
- Multi-task heads (regression + classification per target)
- Loss functions (Tobit, focal, etc.) — applied per-horizon per-cell
- Curriculum sampling strategy
- Evaluation metrics (CRPS, Brier, MCR) — computed per-step
- PredictionFrame output format

### Temporal consistency concern

In an autoregressive model, temporal consistency is implicit — the prediction at t+2 is conditioned on the prediction at t+1, enforcing causal ordering. In direct multi-horizon, predictions at different horizons are independent given the input. The model might predict high conflict at t+3 and low conflict at t+2 and t+4, creating temporal artifacts.

**Mitigations:**
- Temporal smoothness regularization (penalize large changes between consecutive horizons)
- Shared temporal encoder that learns horizon-dependent representations
- Temporal attention (cross-attention between horizon predictions)

---

### Phase 1: Research and design (2-3 weeks)

1. Survey direct multi-horizon architectures compatible with spatial (UNet) structure
2. Design encoder-decoder split: spatial UNet encoder + temporal decoder
3. Decide on horizon representation: shared decoder with horizon embedding vs separate decoders
4. Design temporal consistency mechanism
5. Write ADR documenting architecture decision

### Phase 2: Core implementation (4-6 weeks)

1. Implement temporal encoder (replace ConvLSTM)
2. Implement multi-horizon decoder
3. Modify training loop: process all horizons per window
4. Modify loss computation: aggregate across horizons
5. Modify inference: single forward pass producing all horizons

### Phase 3: Integration (2-3 weeks)

1. Update `HydraNetInference` to support single-pass multi-horizon output
2. Update `InferenceOrchestrator` for new output shape
3. Update `VolumeHandler` for multi-horizon volume construction
4. Update evaluation pipeline (per-horizon metrics)
5. Ensure PredictionFrame compatibility

### Phase 4: Validation (2-3 weeks)

1. Compare against AR baseline on all metrics
2. Per-horizon CRPS comparison (key test: is H=36 quality substantially better?)
3. Temporal consistency analysis
4. Computational cost comparison (training and inference)
5. Uncertainty calibration comparison (MC-Dropout single-pass vs multi-pass)

---

## Decision Criteria

Pursue this path if ANY of the following are true after completing Paths A and E:

1. **Long-horizon degradation persists:** CRPS at step 36 is >3× worse than step 1, even with teacher forcing (Path E). The autoregressive loop is the bottleneck.
2. **Inference cost is prohibitive:** 36 sequential forward passes (× N posterior samples) is too slow for operational deployment. Single-pass inference would be 36× faster.
3. **DynAttn comparison is unfavorable:** If the VIEWS team publishes DynAttn results that substantially outperform our AR model at long horizons, the architecture difference (not the loss function) is likely the cause.
4. **ZINB is adopted (Path C):** If we adopt ZINB (Path C), the architecture is already changing — adding direct MH at the same time captures the full DynAttn approach at low marginal cost.

---

### Arguments for

- Eliminates autoregressive instability **by construction** — no error compounding possible
- Faster inference (1 forward pass vs 36)
- Better long-horizon quality (DynAttn demonstrates this on our data)
- Cleaner training (no teacher forcing curriculum needed)

### Arguments against

- **Massive implementation effort** — months of work, touching every component
- **Loss of temporal causal structure** — must be explicitly regularized
- **Unproven with our architecture** — DynAttn uses transformers, not UNet-LSTM
- **Risk of regression** — a new architecture may lose gains the current model has at short horizons
- **Hidden state utility** — the ConvLSTM's learned hidden state captures long-range temporal dependencies that a direct model must learn differently

---

## Estimated Effort

10-16 weeks for full implementation and validation. This is a multi-sprint initiative, not a feature branch.

---

## Acceptance Criteria

- [ ] Decision documented: pursue direct MH vs keep AR (with evidence from Paths A+E)
- [ ] If pursued: Architecture design ADR written and approved
- [ ] If pursued: Temporal encoder + multi-horizon decoder implemented
- [ ] If pursued: Training pipeline modified for parallel horizon processing
- [ ] If pursued: Inference pipeline produces all horizons in single pass
- [ ] If pursued: Per-horizon CRPS comparison shows improvement at long horizons
- [ ] If pursued: Temporal consistency analysis shows no degradation vs AR
- [ ] If pursued: All existing evaluation metrics computable on new output format

**Comment by Polichinel** [comment]
**PARKED** — deferred fallback (not rejected); see ZINB epic #97 / checklist #104 / C-139. Revisit only if the ZINB head fails. Rationale: loss/calibration investigate-X — deferred; ZINB is the chosen direction
